File: Assignment 2/sorting.hs
{-
  Writing a sorting algorithm that can handle any data type 
  Then test the algorithm with test cases
    - sort numbers ascending by numerical value,
    - sort people alphabetically (lexicographically) by name, and to
    - sort people descending by age, where people of the same age should be sorted alphabetically (lexicographically).
-}

import Data.List (sortBy)

--Allows us to create a data type called Person with the parameters of name and age
data Person = Person { name :: String, age :: Int } deriving (Show, Eq, Ord)

numbers :: [Float]
numbers = [645.41, 37.59, 76.41, 5.31, -34.23, 1.11, 1.10, 23.46, 635.47, -876.32, 467.83, 62.25]

-- Assigns the variable people to the type Person
people :: [Person]
--Creates an array of people
people = [Person "Hal" 20, Person "Susann" 31, Person "Dwight" 19, Person "Kassandra" 21, Person "Lawrence" 25,
          Person "Cindy" 22, Person "Cory" 27, Person "Mac" 19, Person "Romana" 27, Person "Doretha" 32,
          Person "Danna" 20, Person "Zara" 23, Person "Rosalyn" 26, Person "Risa" 24, Person "Benny" 28, Person "Juan" 33,
          Person "Natalie" 25]

main :: IO ()
main = do
    -- Sorts by ascending order by comparing
    let sortedNumbers = sortBy (\a b -> compare a b) numbers 
    putStrLn "Sorted numbers (ascending):"
    print sortedNumbers
    putStrLn ""

    -- Sorts by comparing names
    let sortedPeopleByName = sortBy (\a b -> compare (name a) (name b)) people 
    putStrLn "Sorted people (alphabetically by name):"
    print sortedPeopleByName
    putStrLn ""

    -- Sorts by alphabetical order after checking if ages are the same
    let sortedPeopleByAgeDesc = sortBy (\a b -> if age a /= age b then compare (age b) (age a) else compare (name a) (name b)) people 
    putStrLn "Sorted people (descending by age, then alphabetically by name):"
    print sortedPeopleByAgeDesc

File: Assignment 2/sorting.cs
/*
  Writing a sorting algorithm that can handle any data type 
  Then test the algorithm with test cases
    - sort numbers ascending by numerical value,
    - sort people alphabetically (lexicographically) by name, and to
    - sort people descending by age, where people of the same age should be sorted alphabetically (lexicographically).
*/

using System;
using System.Collections.Generic;

// Class that serves to function as a person
class Person {
    public string Name { 
        get; 
        set; 
    }
    public int Age { 
        get; 
        set; 
    }
}

class Program {
    //Functions as a bubble sort
    static void sort<T>(T[] array, Func<T, T, bool> comp){
        int len = array.Length;
        for (int i = 0; i < len - 1; i++){
            for (int j = 0; j < len - i - 1; j++){
                if (!comp(array[j], array[j + 1])){
                    T index = array[j];
                  array[j] = array[j + 1];
                  array[j + 1] = index;
                } } } }
    static void Main(string[] args){
      //establishes the sample data for the test
      Person[] names = {
          new Person { Name = "Hal", Age = 20 }, new Person { Name = "Susann", Age = 31 }, new Person { Name = "Dwight", Age = 19 }, new Person { Name = "Kassandra", Age = 21 },
          new Person { Name = "Lawrence", Age = 25 }, new Person { Name = "Cindy", Age = 22 }, new Person { Name = "Cory", Age = 27 }, new Person { Name = "Mac", Age = 19 },
          new Person { Name = "Romana", Age = 27 }, new Person { Name = "Doretha", Age = 32 }, new Person { Name = "Danna", Age = 20 }, new Person { Name = "Zara", Age = 23 },
          new Person { Name = "Rosalyn", Age = 26 }, new Person { Name = "Risa", Age = 24 }, new Person { Name = "Benny", Age = 28 }, new Person { Name = "Juan", Age = 33 },
          new Person { Name = "Natalie", Age = 25 }
      };
      double[] num = { 645.41, 37.59, 76.41, 5.31, -34.23, 1.11, 1.10, 23.46, 635.47, -876.32, 467.83, 62.25 };

        // Sort numbers ascending by numerical value
      sort(num, (one, two) => one < two);
      //prints numbers in ascending order
        Console.WriteLine("Sorted numbers (ascending):");
        Console.WriteLine(string.Join(",", num));

      // Sort people alphabetically (lexicographically) by name
      sort(names, (one, two) => string.Compare(one.Name, two.Name) < 0);
      //prints out names in alphabetical order
      Console.WriteLine("Sorted people (alphabetically by name):");
      foreach (var person in names){
          Console.Write($"{person.Name}, {person.Age} ");
          Console.WriteLine();
      }
      Console.WriteLine();
      // sort people descending by age, where people of the same age should be sorted alphabetically (lexicographically)
      sort(names, (a, b) => a.Age > b.Age || (a.Age == b.Age && string.Compare(a.Name, b.Name) > 0));
      //prints ages in numerical order and in alphabetically in the case of people with the same age
      Console.WriteLine("Sorted people (descending by age, then alphabetically by name):");
      foreach (var person in names){
          Console.Write($"{person.Name}, {person.Age} ");
          Console.WriteLine();
      }
  }
}